juju/framework.py

```diff
@@ -552,9 +587,13 @@ def __init__(self, parent, attr_name):
         except NoSnapshotError:
             data = StoredStateData(parent, attr_name)
 
+        # __dict__ is used to avoid infinite recursion.
         self.__dict__["_data"] = data
         self.__dict__["_attr_name"] = attr_name
 
+        # Has to be here because StoredStateData might be restored from persistent storage and it's __init__ will not be invoked.
+        parent.framework.observe(parent.framework.on.pre_commit, self._data)
```

Comment:
As I mentioned in our call, observers are not currently being garbage collected. It's okay to get this in as-is now, but let's please fix that issue on the follow up.